_drafts/2023-08-15-debugging-android-devices.md

```diff
@@ -0,0 +1,274 @@
+---
+title: Debugging Android Devices
+description:
+  What should I include here... <CHRIS COME BACK AND FIX THIS!!>
+author: chayes
+---
+
+All hardware devices in the field experience bugs and need debugging. Android devices in specific are exceptionally complex with several hundred gigabytes of source code, dozens of components, and a wide range of uses.
+
+<!-- excerpt start -->
+
+In this article we will explore the different facilities and tools available to debug Android-based devices and produce robust systems that can handle a wide range of applications from smart fridges, to payment terminals, and of course mobile phones.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## Logging
+
+The first place any debugging starts is usually looking through logs. Android has two sets of logs, Logcat and Kernel logs.
+
+### Logcat
+
+Logcat logs are pretty straightforward, most Android app devs know about logcat There are circular buffers for main, system, crash, radio, and event logs for binaries running within the Android runtime. This can be system services, applications, and native binaries. The CLI allows you to filter specific packages and control the verbosity on a per-package basis.
```

Comment:
```suggestion
Logcat logs are pretty straightforward, most Android app devs know about logcat. There are circular buffers for main, system, crash, radio, and event logs for binaries running within the Android runtime. This can be system services, applications, and native binaries. The CLI allows you to filter specific packages and control the verbosity on a per-package basis. ```

> Logcat logs are pretty straightforward

In the spirit of reaching non-Android audiences, we probably should use a less judgmental voice here, something more like "Logcat logs are the main and most common form of logging on Android devices".

> There are circular buffers for main, system, crash, radio, and event logs

Maybe it'd help to link these buffers to `/dev/log/main` etc as well, since we establish doing that for kernel logs too.

> This can be system services, applications, and native binaries.

It's a bit unclear what `this` is referring to, binaries? buffers? Maybe merge this with the previous line for continuity. 

> The CLI allows you to filter specific packages and control the verbosity on a per-package basis.

We kind of just start talking about a CLI without introducing it. i.e. "Every Android device has a `logcat` CLI, which allows filtering by ..."

Reply:
Appreciate the feedback, and you're definitely right about the tone. This was how I worded it in my conference talk and it doesn't translate well. I'll make the edits you suggested.